include/revng/Pipeline/Pipe.h

```diff
@@ -0,0 +1,311 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <memory>
+#include <set>
+#include <string>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/ADT/iterator_range.h"
+#include "llvm/Support/Error.h"
+
+#include "revng/ADT/STLExtras.h"
+#include "revng/Pipeline/ContainerSet.h"
+#include "revng/Pipeline/Context.h"
+#include "revng/Pipeline/InputOutputContract.h"
+#include "revng/Support/Debug.h"
+#include "revng/Support/FunctionIntrospection.h"
+
+/// A pipe is a class with the following characteristics
+/// * It must have a static constexpr field named Name that is a string
+///   describing its name. Mostly used for debug purposes.
+///
+/// * A std::array<AtomicContract, X> getContract() const method that returns
+///   the contract of such class.
+///
+/// * a void run(T...) method where the first argument type can optionally be
+///   Context and every other type must be the most derived type of a Container.
+///   The operation performed by run MUST be consisted with the contract.
+///
+///
```

Comment:
This sounds like a `concept`!

Reply:
yeah i wanted to do that, the issue is the last requirement. i am not sure how to check that run is a non variadic function with N arguments where the first argument can optionally be context

Reply:
You can make a concept for the first two, and two other concepts for the two options for the third one.
Then make another concept `ConecptForFirstAndSecond and (ConceptForThird_A or ConceptForThird_B)`.

But see [my other comment](https://github.com/revng/revng/pull/55#discussion_r765475568) first.
Maybe the `run` method is becoming to magical.

Reply:
yeah, it is intended to be magical, half of the design is to make sure that it's really easy to write.